src/main/java/net/dv8tion/jda/core/entities/impl/JDAImpl.java

```diff
@@ -122,23 +123,89 @@ public JDAImpl(AccountType accountType, OkHttpClient.Builder httpClientBuilder,
         this.jdaBot = accountType == AccountType.BOT ? new JDABotImpl(this) : null;
     }
 
-    public void login(String token, ShardInfo shardInfo, SessionReconnectQueue reconnectQueue) throws LoginException, RateLimitedException
+    public int login(String token, ShardInfoImpl shardInfo, SessionReconnectQueue reconnectQueue) throws LoginException, RateLimitedException
     {
+        this.shardInfo = shardInfo;
+
         setStatus(Status.LOGGING_IN);
         if (token == null || token.isEmpty())
             throw new LoginException("Provided token was null or empty!");
 
         setToken(token);
         verifyToken();
-        this.shardInfo = shardInfo;
         LOG.info("Login Successful!");
 
+        Pair<String, Integer> gateway = getGatewayAndRecommendedShardCount();
+
+        if (gateway == null) 
+        { 
+            throw new RuntimeException("Could not fetch WS-Gateway!"); 
```

Comment:
Can we use a better exception than `RuntimeException`

Reply:
as `gateway` actually is impossible to be `null` it is pretty useless to add a custom exception for it

Reply:
It doesn't need to be a custom exception, just RuntimeException is considered a "raw exception" which is not a very high code quality